logic/tool/TypeInference.java

```diff
@@ -416,6 +419,16 @@ private void registerIs(TypeVariable resolver, IsConstraint isConstraint) {
             traversal.equalTypes(resolver.id(), register(isConstraint.variable()).id());
         }
 
+        /**
+         * We only extract the type of the IID, and include this information in the type inference
+         * We only look at the type prefix because the static type checker (this class) only read schema, not data
+         */
+        private void registerIID(TypeVariable resolver, IIDConstraint constraint) {
+            TypeVertex type = graphMgr.schema().convert(VertexIID.Thing.of(constraint.iid()).type());
+            if (type == null) throw TypeDBException.of(UNSATISFIABLE_PATTERN, conjunction, constraint);
+            restrict(resolver.id(), iterate(type));
+        }
```

Comment:
restrict the allowed types of a variable if we have an IID by extracting the type of the IID